File: flutter_app/lib/models/match.dart
import 'package:flutter/material.dart';
import 'package:intl/intl.dart';
import '../config/app_config.dart';
import '../config/app_theme.dart';
import 'user.dart';

enum MatchStatus {
  scheduled,
  ongoing,
  completed,
  cancelled;

  static MatchStatus fromString(String? status) {
    switch (status?.toUpperCase()) {
      case 'ONGOING':
        return MatchStatus.ongoing;
      case 'COMPLETED':
        return MatchStatus.completed;
      case 'CANCELLED':
        return MatchStatus.cancelled;
      default:
        return MatchStatus.scheduled;
    }
  }

  String get displayName {
    switch (this) {
      case MatchStatus.scheduled:
        return 'Запланирован';
      case MatchStatus.ongoing:
        return 'Идёт';
      case MatchStatus.completed:
        return 'Завершён';
      case MatchStatus.cancelled:
        return 'Отменён';
    }
  }
}

class Team {
  final String id;
  final String name;
  final String? city;
  final String? logoUrl;
  final bool isHome;

  Team({
    required this.id,
    required this.name,
    this.city,
    this.logoUrl,
    this.isHome = false,
  });

  String? get logoFullUrl {
    if (logoUrl == null || logoUrl!.isEmpty) return null;
    if (logoUrl!.startsWith('http')) return logoUrl;
    return '${AppConfig.uploadsBaseUrl}/$logoUrl';
  }

  factory Team.fromJson(Map<String, dynamic> json) {
    return Team(
      id: json['id'] ?? '',
      name: json['name'] ?? '',
      city: json['city'],
      logoUrl: json['logoUrl'] ?? json['logo_url'],
      isHome: json['isHome'] ?? json['is_home'] ?? false,
    );
  }
}

class Arena {
  final String id;
  final String name;
  final String city;
  final String? address;

  Arena({
    required this.id,
    required this.name,
    required this.city,
    this.address,
  });

  factory Arena.fromJson(Map<String, dynamic> json) {
    return Arena(
      id: json['id'] ?? '',
      name: json['name'] ?? '',
      city: json['city'] ?? '',
      address: json['address'],
    );
  }
}

class FanSector {
  final String id;
  final String name;
  final int capacity;

  FanSector({
    required this.id,
    required this.name,
    required this.capacity,
  });

  factory FanSector.fromJson(Map<String, dynamic> json) {
    return FanSector(
      id: json['id'] ?? '',
      name: json['name'] ?? '',
      capacity: json['capacity'] ?? 0,
    );
  }
}

class Match {
  final String id;
  final Team homeTeam;
  final Team awayTeam;
  final Arena arena;
  final DateTime scheduledAt;
  final DateTime bookingOpenAt;
  final DateTime bookingCloseAt;
  final String? description;
  final bool isHome;
  final MatchStatus status;
  final int cancelDeadlineHours;
  final FanSector? fanSector;
  final int confirmedBookingsCount;
  final int pendingBookingsCount;

  Match({
    required this.id,
    required this.homeTeam,
    required this.awayTeam,
    required this.arena,
    required this.scheduledAt,
    required this.bookingOpenAt,
    required this.bookingCloseAt,
    this.description,
    this.isHome = true,
    required this.status,
    this.cancelDeadlineHours = 24,
    this.fanSector,
    this.confirmedBookingsCount = 0,
    this.pendingBookingsCount = 0,
  });

  String get title => '${homeTeam.name} — ${awayTeam.name}';

  int get availableSeats {
    if (fanSector == null) return 0;
    final left = fanSector!.capacity - confirmedBookingsCount;
    return left > 0 ? left : 0;
  }

  bool get isBookingOpen {
    final now = DateTime.now();
    if (status == MatchStatus.completed || status == MatchStatus.cancelled) {
      return false;
    }
    if (now.isBefore(bookingOpenAt) || now.isAfter(bookingCloseAt)) {
      return false;
    }
    return availableSeats > 0;
  }

  String get formattedDateTime {
    return DateFormat("d MMMM 'в' HH:mm", 'ru').format(scheduledAt);
  }

  String get formattedDate {
    return DateFormat("d MMMM", 'ru').format(scheduledAt);
  }

  String get formattedTime {
    return DateFormat("HH:mm").format(scheduledAt);
  }

  double get occupancyPercent {
    if (fanSector == null || fanSector!.capacity == 0) return 0.0;
    return (confirmedBookingsCount / fanSector!.capacity).clamp(0.0, 1.0);
  }

  factory Match.fromJson(Map<String, dynamic> json) {
    return Match(
      id: json['id'] ?? '',
      homeTeam: Team.fromJson(json['homeTeam'] ?? json['home_team'] ?? {'name': 'Хозяева'}),
      awayTeam: Team.fromJson(json['awayTeam'] ?? json['away_team'] ?? {'name': 'Гости'}),
      arena: Arena.fromJson(json['arena'] ?? {'name': 'Арена', 'city': 'Город'}),
      scheduledAt: DateTime.tryParse(json['scheduledAt'] ?? '') ?? DateTime.now(),
      bookingOpenAt: DateTime.tryParse(json['bookingOpenAt'] ?? '') ?? DateTime.now(),
      bookingCloseAt: DateTime.tryParse(json['bookingCloseAt'] ?? '') ?? DateTime.now(),
      description: json['description'],
      isHome: json['isHome'] ?? json['is_home'] ?? true,
      status: MatchStatus.fromString(json['status']),
      cancelDeadlineHours: json['cancelDeadlineHours'] ?? 24,
      fanSector: json['fanSector'] != null ? FanSector.fromJson(json['fanSector']) : null,
      confirmedBookingsCount: json['confirmedBookingsCount'] ?? json['confirmed_bookings_count'] ?? 0,
      pendingBookingsCount: json['pendingBookingsCount'] ?? json['pending_bookings_count'] ?? 0,
    );
  }
}

class MatchParticipant {
  final String id;
  final User user;
  final String status;
  final String attendanceStatus;

  MatchParticipant({
    required this.id,
    required this.user,
    required this.status,
    required this.attendanceStatus,
  });

  factory MatchParticipant.fromJson(Map<String, dynamic> json) {
    return MatchParticipant(
      id: json['id'] ?? '',
      user: User.fromJson(json['user'] ?? {}),
      status: json['status'] ?? 'PENDING',
      attendanceStatus: json['attendance']?['status'] ?? 'UNKNOWN',
    );
  }
}

File: ios/FanZone/FanZone/Models/Match.swift
import Foundation
import SwiftUI

// MARK: - Match Status

enum MatchStatus: String, Codable, CaseIterable {
    case scheduled = "SCHEDULED"
    case ongoing   = "ONGOING"
    case completed = "COMPLETED"
    case cancelled = "CANCELLED"

    var displayName: String {
        switch self {
        case .scheduled: return "Запланирован"
        case .ongoing:   return "Идёт"
        case .completed: return "Завершён"
        case .cancelled: return "Отменён"
        }
    }
}

// MARK: - Booking Window Status (derived)

enum BookingWindowStatus {
    case open
    case almostFull(available: Int)
    case closed
    case completed

    var displayName: String {
        switch self {
        case .open:                 return "Бронирование открыто"
        case .almostFull(let n):    return "Мест мало (\(n))"
        case .closed:               return "Бронирование закрыто"
        case .completed:            return "Матч завершён"
        }
    }

    var emoji: String {
        switch self {
        case .open:       return "🟢"
        case .almostFull: return "🟡"
        case .closed:     return "🔴"
        case .completed:  return "⚪"
        }
    }

    var color: Color {
        switch self {
        case .open:       return AppTheme.success
        case .almostFull: return AppTheme.warning
        case .closed:     return AppTheme.error
        case .completed:  return AppTheme.neutral
        }
    }
}

// MARK: - Team

struct Team: Codable, Identifiable, Equatable {
    let id: String
    let name: String
    let city: String?
    let logoUrl: String?
    let isHome: Bool?

    var logoURL: URL? {
        guard let path = logoUrl, !path.isEmpty else { return nil }
        if path.hasPrefix("http") { return URL(string: path) }
        return AppConfiguration.uploadsBaseURL.appendingPathComponent(path)
    }
}

// MARK: - Arena

struct Arena: Codable, Identifiable, Equatable {
    let id: String
    let name: String
    let city: String
    let address: String?
}

// MARK: - Fan Sector

struct FanSector: Codable, Identifiable, Equatable {
    let id: String
    let matchId: String?
    let name: String
    let capacity: Int
}

// MARK: - Match

struct Match: Codable, Identifiable, Equatable {
    let id: String
    let homeTeam: Team
    let awayTeam: Team
    let arena: Arena
    let scheduledAt: Date
    let bookingOpenAt: Date
    let bookingCloseAt: Date
    let description: String?
    let isHome: Bool
    let status: MatchStatus
    let cancelDeadlineHours: Int?
    let fanSector: FanSector?
    let confirmedBookingsCount: Int?
    let pendingBookingsCount: Int?
    let createdAt: Date?

    static func == (lhs: Match, rhs: Match) -> Bool { lhs.id == rhs.id }

    // MARK: Computed

    var title: String { "\(homeTeam.name) — \(awayTeam.name)" }

    var isUpcoming: Bool { scheduledAt > Date() && status == .scheduled }
    var isPast: Bool     { scheduledAt < Date() || status == .completed }

    var availableSeats: Int {
        guard let sector = fanSector, let confirmed = confirmedBookingsCount else { return 0 }
        return max(0, sector.capacity - confirmed)
    }

    var bookingWindowStatus: BookingWindowStatus {
        let now = Date()
        if status == .completed || status == .cancelled { return .completed }
        if now < bookingOpenAt || now > bookingCloseAt  { return .closed }
        let available = availableSeats
        if available == 0 { return .closed }
        if available <= 10 { return .almostFull(available: available) }
        return .open
    }

    var isBookingOpen: Bool {
        if case .open       = bookingWindowStatus { return true }
        if case .almostFull = bookingWindowStatus { return true }
        return false
    }

    var formattedSchedule: String {
        scheduledAt.ruDateTimeString
    }

    var occupancyPercent: Double {
        guard let sector = fanSector, sector.capacity > 0,
              let confirmed = confirmedBookingsCount else { return 0 }
        return Double(confirmed) / Double(sector.capacity)
    }
}

// MARK: - Match Participant (for manager view)

struct MatchParticipant: Codable, Identifiable {
    let id: String          // booking id
    let user: User
    let status: BookingStatus
    let attendance: AttendanceRecord?
    let createdAt: Date
}
